File: Baekjoon/yerim/퇴사.cpp
// [14501] 퇴사
// dp[i] = i번째 날부터 마지막날까지의 최대 이익
#include <iostream>
using namespace std;

int n, t[16], p[16], dp[16];

int max(int a, int b)
{
    if(a > b) return a;
    else return b;
}

void getMaxProfit()
{   
    dp[n] = (t[n] == 1) ? p[n] : 0;
    for(int i = n-1; i >= 1; i--) {
        if(i + t[i] <= n + 1) {
            dp[i] = max(dp[i+1], p[i] + dp[i + t[i]]);
        }
        else {
            dp[i] = dp[i+1];
        }
    }
}

void input()
{
    cin >> n;
    for(int i = 1; i <= n; i++) {
        cin >> t[i] >> p[i];
    }
}

int main(void)
{
    input();
    getMaxProfit();
    cout << dp[1] << '\n';
}

File: Baekjoon/chulmin/퇴사.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class 퇴사 {
	static int N;
	static int[] dp;
	static ArrayList<int[]>[] list;
	
	public static void main(String[] args) throws IOException {
		input();
		pro();
	}

	private static void pro() {
		
		for(int i=1; i<=N+1; i++) {
			dp[i] = dp[i-1];
			
			for(int[] j : list[i]) {
				int cur = dp[i - j[0]] + j[1];
				dp[i] = Math.max(cur, dp[i]);
			}
		}
//		for(int i: dp) {
//			System.out.print(i + " ");
//		}System.out.println();
		System.out.println(dp[N+1]);
	}

	private static void input() throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		
		N = Integer.parseInt(br.readLine());
		list = new ArrayList[N + 6];
		dp = new int[N+2];
		
		for(int i=0; i<=N+5; i++) {
			list[i] = new ArrayList<int[]>();
		}
		
		for(int i=1; i<=N; i++) {
			st = new StringTokenizer(br.readLine());
			int t = Integer.parseInt(st.nextToken());
			int p = Integer.parseInt(st.nextToken());
			
			list[i+t].add(new int[] {t, p});
		}
		
		br.close();
	}
}
